text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,162 @@
+
+Having the `BOM` be a part of the layer content metadata also means that we can pin-point which layer contributed a BOM entry without having to implement extra logic in downstream buildpacks.
+
+- What is the expected outcome?
+
+The lifecycle is able to generate the final `BOM` by combining the `BOM` entries from top level metadata files like `launch.toml` and `build.toml` and also the ones from `<layer-name>.toml` which handling layer caching correctly i.e. if the layer is cached and re-used in the new build - the `BOM` associated with that layer will also be re-used.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The layer content metadata files would be updated to include - 
+
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+# arbitrary metadata describing the dependency
+```
+
+For each dependency contributed to the app image or build environment, the buildpack layer:
+
+- SHOULD add a bill-of-materials entry to the bom array describing the dependency, where:
+  - name is REQUIRED.
+  - metadata MAY contain additional data describing the dependency.
+  
+The buildpack SHOULD NOT add bom to a layer content metadata file describing the contents that are not contributed by the same layer. `BOM` entries that describe contribution to the application directory should go in the buildpack-level `BOM` entries in `launch.toml` or `build.toml`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle would be responsible for the generation of the appropriate metadata files and labels given partial `BOM` entries in the `launch.toml`, `build.toml` and `<layer-name>.toml`. The generation of these artifacts given `BOM` entries in `<layer-name>.toml` would be controlled as follows - 
+
+- If the layer has `types.launch` set to `true` then all the `bom` entries will be contributed to the app image in `io.buildpacks.build.metadata` and `<layers>/config/metadata.toml`.
+- If the layer has `types.build` set to `true` then all the `bom` entries will be written to `report.toml`.
+- If the layer has `types.cache` set to `true` and if the layer is reused in the next build, the respective `bom` entries should be re-populated from the cached layer content metadata file.
+- If `types.cache` is set to `false`, but `types.launch` is set to `true` before a rebuild, the layer content metdata file should be restored with the `BOM` section. If the buildpack sets `types.launch` to `true` after the re-build, the final value of `bom` after the re-build should be used.
+- If neither `types.launch` or `types.build` are set to `true` but the layer is present during the export phase the `bom` entries should STILL be written to `report.toml` as the contents of this layer may have been used to generate other artifacts in the same buildpack and someone went through the effort of adding these entries to the `bom`. 
```

Comment:
I feel that this slightly contradicts the spirit of https://github.com/buildpacks/rfcs/blob/main/text/0052-opt-in-layer-caching.md in that 0052 made it so that buildpacks by "doing nothing" don't get layer re-use. Here, if a buildpack does nothing (doesn't set `types` but also doesn't delete the layer.toml file) the layer is still "there" during export and would be included in the BOM even if it really didn't contribute to the build at all. 

This is sort of a tough spot because we can't really know if a buildpack did use a layer in the **current** build that was restored from a previous build.

Reply:
Summarizing from WG: adding to the build BOM is the right thing to do because even if a layer wasn't actually used, its presence in the build container could still be noteworthy. I think that satisfies the concern :)